Local file storage extension

djaglowski opened the pull request:
Follows #2883 

Will resolve #2287

### Summary

This PR includes:
- A full implementation of a `file_storage` extension, which can read and write data to the local file system. Any component in the collector may make use of this extension.
- Updates to `stanza/internal` to allow stanza-based receivers to use the extension for checkpoints.
- A new testbed scenario that has the filelogreceiver using the extension

Configuration of the extension is simple. 
```yaml
  file_storage:
  file_storage/all_settings:
    directory: /var/lib/otelcol/mydir
    timeout: 2s
```

The extension is made available to component's via the `host` parameter in their `Start` method:
```go
func (r *receiver) Start(ctx context.Context, host component.Host) error {
	for _, ext := range host.GetExtensions() {
		if se, ok := ext.(storage.Extension); ok {
			client, err := se.GetClient(ctx, component.KindReceiver, r.NamedEntity)
			if err != nil {
				return err
			}
			r.storageClient = client
			return nil
		}
	}
	r.storageClient = storage.NewNopClient()
        ...
}
```

djaglowski committed:
Add local file storage extension

djaglowski commented:
This PR is ready for review, with a couple caveats:

- Unit test coverage is below the threshold. Remaining cases are mostly very specific error scenarios. Will add more tests.

- The `lint` check is failing in a way that I have not been able to reproduce yet. Will continue to investigate.
```
make -C ./testbed lint
make[1]: Entering directory '/home/circleci/project/testbed'
golangci-lint run --allow-parallel-runners
ERRO Running error: context loading failed: no go files to analyze 
make[1]: *** [../Makefile.Common:93: lint] Error 5
make[1]: Leaving directory '/home/circleci/project/testbed'
make: *** [Makefile:143: for-all-target-./testbed] Error 2
```

tigrannajaryan commented:
Please rebase.

tigrannajaryan commented on extension/storage/local/filestorage/client.go:
```diff
@@ -0,0 +1,96 @@
+	db *bbolt.DB
+}
+
+func newClient(file string, timeout time.Duration) (*fileStorageClient, error) {
```
Is `file` the file name? File path?

tigrannajaryan replied:
What does `timeout` do?

djaglowski replied:
`file` is the full file path. I'll rename this accordingly.

`timeout` sets the maximum time to wait when trying to obtain a lock on the file. Without setting this, it can wait indefinitely. I'm not certain that this needs to be configurable, but my thinking is that it shouldn't require a rebuild in order to change it, if it ever proves to be an issue. I've also added a note to the documentation.

tigrannajaryan commented on extension/storage/local/filestorage/client.go:
```diff
@@ -0,0 +1,96 @@
+
+func newClient(file string, timeout time.Duration) (*fileStorageClient, error) {
+	options := &bbolt.Options{Timeout: timeout}
+	db, err := bbolt.Open(file, 0666, options)
```
I don't remember the meaning of 0666 from the top of my head. Is it read/write, non-executable?

djaglowski replied:
That's correct. This is probably too permissive even. I believe 0600 is sufficient, so I'll change it to that.

tigrannajaryan commented on extension/storage/local/filestorage/client.go:
```diff
@@ -0,0 +1,96 @@
+	get := func(tx *bbolt.Tx) error {
+		bucket := tx.Bucket(defaultBucket)
+		if bucket == nil {
+			return fmt.Errorf("storage not initialized")
```
Nit (I think our linter wants this):
```suggestion
			return errors.New("storage not initialized")
```

tigrannajaryan commented on extension/storage/local/filestorage/client.go:
```diff
@@ -0,0 +1,96 @@
+	return c.db.Update(set)
+}
+
+// Delete will delete data associated with the specified key
```
Does this return an error if the key does not exist?

djaglowski replied:
The short answer is no.

The longer version: I think that all methods on this interface should only return an error if a "real" problem occurred (such as the storage not being available). 

There are a couple reasons why it makes more sense this way, but of course am open to feedback:
1. I think we need a way to differentiate between data operations [overwrite | not-found | no-op] from "real" problems.
2. Since we're working on a key/value store in Go, I think it makes sense to mirror the behavior of a Golang `map`. This means `Set` doesn't error if key already exists - it just overwrites the value. `Get` doesn't error if a key is not found - it just returns `nil`. `Delete` doesn't error if the key doesn't exist - it just no-ops.

That said, I think there is a use case for returning an indication of whether or not anything was deleted. However, one could call `Get` immediately before `Delete` if this needs to be ascertained.

tigrannajaryan replied:
I agree with all you said. Please document in the code.

tigrannajaryan commented on extension/storage/local/filestorage/default_others.go:
```diff
@@ -0,0 +1,22 @@
+
+// TODO do we like these path names?
+func getDefaultDirectory() string {
+	return "/var/lib/otelcol/file_storage"
```
What about:
```suggestion
	return "/var/lib/otelcol/data"
```
Do you want to see the extension name in the path?

djaglowski replied:
I don't have a strong opinion on this one, but one possible reason to indicate the extension type in the path is to allow for a degree of isolation from alternate implementations.

Perhaps a file naming extension would protect against this. (i.e. `/var/lib/otelcol/data/kind_type_name.otdb`)

tigrannajaryan replied:
Actually, I think what you have is better. Let's name the directory after the extension.

tigrannajaryan commented on extension/storage/local/filestorage/default_windows.go:
```diff
@@ -0,0 +1,27 @@
+
+// TODO do we like this path?
+func getDefaultDirectory() string {
+	return filepath.Join(os.Getenv("ProgramData"), "Otelcol", "FileStorage")
```
```suggestion
	return filepath.Join(os.Getenv("ProgramData"), "Otelcol", "Data")
```

tigrannajaryan replied:
Disregard this.

tigrannajaryan commented on extension/storage/local/filestorage/extension.go:
```diff
@@ -0,0 +1,106 @@
+func newLocalFileStorage(logger *zap.Logger, config *Config) (component.Extension, error) {
+	info, err := os.Stat(config.Directory)
+	if (err != nil && os.IsNotExist(err)) || !info.IsDir() {
+		return nil, fmt.Errorf("directory must exist: %v", err)
```
Should we try to create it or we rely on installation packages or the user to do it? Probably the later is best.

djaglowski replied:
I prefer the latter as well.

tigrannajaryan commented on extension/storage/local/filestorage/extension.go:
```diff
@@ -0,0 +1,106 @@
+		return nil, fmt.Errorf("directory must exist: %v", err)
+	}
+
+	testFile := filepath.Join(config.Directory, "test.tmp")
+	options := &bbolt.Options{Timeout: time.Second}
+	db, err := bbolt.Open(testFile, 0666, options)
+	if err != nil {
+		return nil, err
+	}
```
Why do we need to test it?

djaglowski replied:
We don't necessarily need to. In fact, the same is done at client creation.

I think I'll remove this. However, one possible reason to keep it is that the error is then associated with the extension, rather than with the component using the extension, which may be more helpful to the user.

tigrannajaryan commented on extension/storage/local/filestorage/extension.go:
```diff
@@ -0,0 +1,106 @@
+
+// GetClient returns a storage client for an individual component
+func (lfs *localFileStorage) GetClient(ctx context.Context, kind component.Kind, ent config.NamedEntity) (storage.Client, error) {
+	rawName := fmt.Sprintf("%s_%s_%s", kindString(kind), ent.Type(), ent.Name())
```
Are Type() and Name() guaranteed to only have characters allowed in file paths?

djaglowski replied:
Looking closer, I see that I misunderstood the next line. `cleanName := filepath.Clean(rawName)`

Are you aware of an idiomatic solution here? I don't see anything obvious from a quick search. Will look into it more though.

tigrannajaryan replied:
The invalid characters need to be encoded using valid ones. I guess something like urlencode/QueryEscape (https://golang.org/pkg/net/url/#QueryEscape) should work. You will need to check that it escape the correct set of characters (it needs to be a superset of what we need). I am not aware of a precise function that does this for path names.

tigrannajaryan commented on extension/storage/local/filestorage/extension.go:
```diff
@@ -0,0 +1,106 @@
+
+// Shutdown will close any open databases
+func (lfs *localFileStorage) Shutdown(context.Context) error {
+	for _, client := range lfs.clients {
```
For the future: here we have an opportunity to to garbage collect files that don't have clients (I am not sure we have a good place to do it at startup).

djaglowski replied:
I like this. A configurable TTL could do this nicely.

```yaml
  file_storage:
    component_data_ttl: 24h
```

```go
func (lfs *localFileStorage) Shutdown(context.Context) error {
    for _, client := range lfs.clients {
        client.close()
    }
    for f := range findStorageFiles(lfs.directory) {
        if f.LastModified() > time.Now() - lfs.componentDataTTL {
            rm(f)
        }
    }
}
```

tigrannajaryan replied:
Sounds good. You can also skip files for which we had clients, no need to delete those. Can be in a future PR.

tigrannajaryan commented on extension/storage/local/filestorage/factory.go:
```diff
@@ -0,0 +1,55 @@
+)
+
+// The value of extension "type" in configuration.
+const typeStr config.Type = "local_file_storage"
```
Looks a bit too long. Maybe simply "file_storage"? Non-local ones likely will have their specific qualifier.

tigrannajaryan commented on internal/stanza/storage.go:
```diff
@@ -0,0 +1,62 @@
+)
+
+func (r *receiver) setStorageClient(ctx context.Context, host component.Host) error {
+	// TODO define expected behavior if >1 storage extension found
```
Probably fail. In the future we can introduce a config setting to select the storage if there is more than one.

tigrannajaryan reviewed (commented):
I like how minimalistic this turned out to be.
Can we add simple perf tests for the storage and publish the results in the PR?

djaglowski commented:
The PR includes one additional testbed scenario that utilizes the new extension. It shows nearly no impact on performance. However, this usage of the extension by the `filelogreceiver` is very light, so this is somewhat expected.
```
Log10kDPS/filelog                       |PASS  |     15s|    17.8|    19.3|         59|         74|    150000|        150000|
Log10kDPS/filelog_checkpoints           |PASS  |     15s|    17.5|    19.3|         60|         73|    149900|        149900|
```

I've added a benchmark test directly to the client, and it unfortunately appears to perform quite poorly:
`BenchmarkClientOperations-8   	      30	  38213851 ns/op	   19442 B/op	     120 allocs/op`

There is a fairly obvious improvement, which I suggest we include in a separate PR. Stanza used a local `map[string][]byte` to cache data in memory, and synced the cache to disk on demand. The `storage.Client` interface does not allow control of the syncing functionality to the components, but I believe we sync 1) periodically on a timer (possibly configurable, but not necessarily), and 2) at `Shutdown`.

tigrannajaryan commented:
> I've added a benchmark test directly to the client, and it unfortunately appears to perform quite poorly:
> `BenchmarkClientOperations-8 30 38213851 ns/op 19442 B/op 120 allocs/op`

This is surprisingly slow, by a few orders of magnitude slower than I would expect and it is not good enough for our other needs (ondisk buffers for exporter queues). On an SSD even if fsyncs every time I would not expect 38ms per operation. Do you know which operation of the 3 (Get/Set/Delete) is slow?

djaglowski commented:
I should have mentioned, that the benchmark was performing all 3 operations per iteration, but obviously this is still very slow. 

I've split them into separate benchmarks, and it appears that each performs similarly.
```
BenchmarkClientGet-8      	      81	  13057659 ns/op	    5923 B/op	      32 allocs/op
BenchmarkClientSet-8      	      87	  12460373 ns/op	    6796 B/op	      45 allocs/op
BenchmarkClientDelete-8   	      82	  12823977 ns/op	    5925 B/op	      32 allocs/op
```

I will try out the short-term local cache idea and report back.

tigrannajaryan commented:
> I will try out the short-term local cache idea and report back.

I am not sure I like this. I'd very much prefer a proper file storage that writes to file immediately (without fsync) so that if we crash the data is still stored. The performance we see is extremely slow and adding a in-memory cache defeats the purpose for high-rate cases.

I think it is OK to merge this PR and ignore the PR for now, but then immediately look for a better KV storage without spending time on adding a cache. I cannot imagine a KV storage being so slow (is it possible we do something wrong?). Surely something else should exist that has more reasonable performance.

tigrannajaryan commented:
OK, from what I see bolt by default fsyncs every transaction. Setting `db.NoSync = true` improves performance by 3 orders of magnitude on my machine. I believe we don't need a fsync.

djaglowski commented:
Thanks for looking into this Tigran. I see similar gains. 
```
BenchmarkClientGet-8      	   96832	     12863 ns/op	    5494 B/op	      31 allocs/op
BenchmarkClientSet-8      	   67626	     15891 ns/op	    6351 B/op	      44 allocs/op
BenchmarkClientDelete-8   	   97971	     12196 ns/op	    5493 B/op	      31 allocs/op
```

Based on the comments on it, I have some reservations about using `NoSync`, but as you said, we can likely find a better library.

tigrannajaryan commented:
> Based on the code comments on `NoSync`, I have some reservations about using it, but as you said, we can likely find a better library.

What are your reservations? From what I see with NoSync=true the data is still written to the file (as expected). The only difference is that there is no longer a fsync at the end of each transaction. This means that we are unprotected from hardware failures and from kernel crashes. We are still protected from our process crashes and restarts. I believe that is a good enough guarantee for what we use the storage for. I think it is acceptable for us to loose data or have corrupted data in case of catastrophic events like hardware failure or in case of kernel crashes.

We can always add a config option in the future for users to enable fsync if they feel they need it, but I think a default of fsync=false is a fine approach.

djaglowski commented:
> What are your reservations?

My reservation primarily comes from the comment no the `DB.NoSync` field, particularly the last line: 

```
type DB struct {
    ...
    // Setting the NoSync flag will cause the database to skip fsync()
    // calls after each commit. This can be useful when bulk loading data
    // into a database and you can restart the bulk load in the event of
    // a system failure or database corruption. Do not set this flag for
    // normal use.
    //
    // If the package global IgnoreNoSync constant is true, this value is
    // ignored.  See the comment on that constant for more details.
    //
    // THIS IS UNSAFE. PLEASE USE WITH CAUTION.
    NoSync bool
```

I think perhaps the author is just being overly cautionary and that I am echoing that out of ignorance on the implementation details.

Your explanation makes sense to me though. I think you probably have a better understanding and higher level of confidence than I on the nuances of file system operations, so I'm happy to defer to your judgement here and have updated the setting accordingly.

I believe I have addressed all feedback, or except what's captured in the following issues: #3148, #3149, #3150, #3152

tigrannajaryan commented:
> I think perhaps the author is just being overly cautionary and that I am echoing that out of ignorance on the implementation details.

I quickly traced through bolt code, obviously I don't know it in details, but I can see that the flag controls whether OS fsync is called or no (as I would expect). The warning in comment looks scary but I do not see anything in the code that contradicts my understanding. I suggest that we go with NoSync=true and we can adjust if needed.

tigrannajaryan commented:
LGTM, please resolve the conflicts and we can merge.

Is it possible to add a test that verifies that checkpoint storage works correctly with filelog? I think we should be able to start/stop/start the receiver and see if the checkpoints are picked up. Can be a separate PR.

djaglowski committed:
Merge branch 'main' into storage-extension-2

djaglowski commented:
Thanks @tigrannajaryan, conflicts resolved.

The first test in `internal/stanza/storage_test.go` cycles the receiver and validates that data has persisted.

djaglowski committed:
Increase CPU threshold on testbed scenario

tigrannajaryan commented:
> The first test in `internal/stanza/storage_test.go` cycles the receiver and validates that data has persisted.

Yes, but it does not verify that the checkpoints functionality actually works, only that a key/value can be stored and retrieved. I want to see an end to end test. Create a log fie, collect part of it, shutdown the receiver, start it again, make sure it continued collection from the right place, nothing was missed, nothing collected twice. Again, no need to add to this PR, but I believe it is very important to have end to end tests like this.

djaglowski commented:
@tigrannajaryan Understood. Will track with #3157

tigrannajaryan reviewed (approved):
Thanks @djaglowski for making a great progress on this.

tigrannajaryan commented:
@pmm-sumo I merged this but if you have any comments, particularly with regards to the usage in the disk buffer please feel free to submit issues.

pmm-sumo commented:
> @pmm-sumo I merged this but if you have any comments, particularly with regards to the usage in the disk buffer please feel free to submit issues.

This code looks great, I have one comment so far. If I read this piece correctly, there should be [a single extension defined at a given time](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/3087/files#diff-abf81ed736f9d7715e280bb4f8b19264d0528a497f4615986dd1d9e5126ac4a9R30-R34). Is that accurate? I was thinking if maybe the components using storage extension should have the ability to refer a specific one if more than one is configured, but I might be missing something here

djaglowski commented:
@pmm-sumo Thanks for taking a close look at it. Your interpretation is correct. The constraint is not necessarily meant to exist long term though - just seemed like a reasonable starting point.
